### ISSUE TITLE
Support vendoring of Cargo dependencies

### ISSUE THREAD

**Issue by AndrewGaspar** [spack/spack#16045]
To support Rust packages on air-gapped machines, it's necessary to support offline builds.

Discussed briefly offline with @adamjstewart 

### Rationale

Rust has a language-centric package manager called Cargo. This package manager is the primary way that Rust packages add third-party dependencies. Practically every Rust project uses at least some third-party dependency, so it's necessary to be able to pull them in some way. By default, Cargo attempts to pull dependencies from the Internet, which breaks offline builds that rely on `spack mirror`. However, Cargo supports vendoring dependencies via `cargo vendor`. Therefore, it should be possible to implement this in a way that keeps the entire process inside the spack ecosystem, and without packages having to jump through hoops.

### Description

Possible Solutions, ordered from most desirable to least desirable:
#### Native Support for Vendoring from a Cargo.lock file
The best possible solution is for Spack to offer some way for a package to opt into vendoring the packages from its Cargo.lock file during the fetch phase. Native support for Cargo is useful because it provides an easy on-ramp for all Rust packages. You could imagine adding a `CargoPackage` base class that automates the entire fetch and build process for packages which use cargo as their top level build system.

Since Cargo.lock checksums each dependency, it would be possible to verify the integrity of each package without having to resort to putting every cargo package into spack.

Ideally, being able to invoke this behavior wouldn't require `CargoPackage` - you can imagine that there are some packages which use Rust internally that would like to be able to vendor their dependencies, but have a different top-level build system. So there should be some way for any package flavor to opt into this.

#### Rust Packages Ship Releases That Vendor Dependencies
Another possibility is to just have every Rust package that Spack supports to ship their own vendored dependencies in their release tar balls. This solves the problem "cleanly" without any involvement from Spack, but it has some downsides:
1) Most existing Rust projects do not do this, so anytime you'd want to add a Rust package, you'd have to request that the project maintainer issue releases with vendored dependencies.
2) It means you can't just point spack at a "crates.io" URL - you'd need specialized releases hosted elsewhere.

#### Support for Package-Overload of the Fetch Stage
Another solution is to just punt it to the package. However, this would still require feature work in Spack to allow a package to add additional steps to "spack fetch". It would also need to be run after the base package is downloaded so that the `Cargo.lock` file could be checked.

This would be maybe acceptable, but it would lead to each Rust package having to roll the fetch behavior themselves.

#### Import All of crates.io into Spack
You could imagine just creating a spack package for every needed Rust crate in a given dependency tree. This is not desirable for a number of reasons:
1) omg, _could_ you imagine?
2) Rust is reluctant to fully support sharing library "crate" builds across separate Rust workspaces, so the build wouldn't really benefit from spack's model of sharing dependencies when possible. The library packages would essentially just be source distributions, and they wouldn't get built until the top level Rust package is built. I think it's technically possible to share crate builds, but it's not the typical workflow.
3) Those library "crates" are not useful to non-Rust projects. They export a Rust ABI and interface - only executables and libraries which export a portable ABI (such as a C-style interface) are useful to other packages in Spack.
4) You're essentially re-writing Cargo, which probably would lead to bugs down the line.

### General information

- [x] Spack Version: `0.14.1-624-1b0dbc09c`
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by AndrewGaspar** [comment]
See https://github.com/spack/spack/pull/16039 for an example of why this would be useful.

**Comment by adamjstewart** [comment]
Also adding @hartzell to the discussion, as the Rust build system seems similar to Go and he is fixing this same problem for Go in #14061.

**Comment by AndrewGaspar** [comment]
Yeah, reading through that PR, it seems like there are a lot of common concerns - especially fetching and vendoring packages from external package managers. I'll try to do a more thorough review of that at some point from the perspective of what Rust needs.
